book/stages/README.md

```diff
@@ -0,0 +1,162 @@
+
+Each stage within the pipeline implements the `Stage` trait which provides function interfaces to get the stage id, execute the stage and unwind the state if there was an issue during the stage execution. 
+
+
+[Filename: crates/stages/src/stage.rs](https://github.com/paradigmxyz/reth/blob/main/crates/stages/src/stage.rs#L64)
+```rust
+#[async_trait]
+pub trait Stage<DB: Database>: Send + Sync {
+    /// Get the ID of the stage.
+    ///
+    /// Stage IDs must be unique.
+    fn id(&self) -> StageId;
+
+    /// Execute the stage.
+    async fn execute(
+        &mut self,
+        db: &mut StageDB<'_, DB>,
+        input: ExecInput,
+    ) -> Result<ExecOutput, StageError>;
+
+    /// Unwind the stage.
+    async fn unwind(
+        &mut self,
+        db: &mut StageDB<'_, DB>,
+        input: UnwindInput,
+    ) -> Result<UnwindOutput, Box<dyn std::error::Error + Send + Sync>>;
+}
+```
+
+To get a better idea of what is happening at each part of the pipeline, lets walk through what is going on under the hood within the `execute()` function at each stage, starting with `HeadersStage`.
+
+<br>
+
+## HeadersStage
+
+The `HeadersStage` is responsible for syncing the block headers, validating the header integrity and writing the headers to the database. When the `execute()` function is called, the head of the chain is updated to the most recent block height previously executed by the stage. At this point, the node status is also updated with the most recent block height, hash and total difficulty. These values are used during any new eth/65 handshakes. After updating the head, a stream is established with other peers in the network to sync the missing chain headers between the most recent state stored in the database and the chain tip. This stage relies on the stream to return the headers in descending order staring from the chain tip down to the latest block in the database.
+
+Each value yielded from the stream is a `SealedHeader`. 
+
+[File: crates/primitives/src/header.rs](https://github.com/paradigmxyz/reth/blob/main/crates/primitives/src/header.rs#L207)
+```rust
+/// A [`Header`] that is sealed at a precalculated hash, use [`SealedHeader::unseal()`] if you want
+/// to modify header.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct SealedHeader {
+    /// Locked Header fields.
+    header: Header,
+    /// Locked Header hash.
+    hash: BlockHash,
+}
+```
+
+Each `SealedHeader` is then validated to ensure that it has the proper parent. Note that this is only a basic response validation, and a full header validation adhering to the consensus spec occurs in the `downloader`. Following the basic header validation, each header is then written to the database. If a header is not valid or the stream encounters any other error, the error is propagated up through the stage execution, the db changes are unwound and the stage is resumed from the most recent valid state.
```

Comment:
validation according to consensus is going to stick out to the reader, and at a glance this might be read as "this stage does not do validation according to consensus". additionally, it's not super clear what the `downloader` is here.

i think it'd be helpful to mention, either here or in the paragraph above, that the stream that is established is an implementation of the `HeaderDownloader` trait (and, I suppose, pull in the definition for that trait), and double-down on the point that the implementation of that trait must validate the downloaded header according to the consensus specification.

so basically we construct a `SealedHeader` for the current latest header, then we request from our peers to download (in reverse order) all headers since the `SealedHeader`, and it is in the implementation of the downloader that consensus validation occurs

Reply:
sorry, i know that was a super verbose comment, just wanted to get it down - please let me know if any part of it could be clarified